lib/concurrent/agent.rb

```diff
@@ -1,260 +1,579 @@
+  # the new value is successfully validated. Observation will *not* occur if the
+  # action raises an exception, if validation fails, or when a {#restart} occurs.
+  #
+  # When notified the observer will receive three arguments: `time`, `old_value`,
+  # and `new_value`. The `time` argument is the time at which the value change
+  # occurred. The `old_value` is the value of the Agent when the action began
+  # processing. The `new_value` is the value to which the Agent was set when the
+  # action completed. Note that `old_value` and `new_value` may be the same.
+  # This is not an error. It simply means that the action returned the same
+  # value.
+  #
+  # ## Nested Actions
+  #
+  # It is possible for an Agent action to post further actions back to itself.
+  # The nested actions will be enqueued normally then processed *after* the
+  # outer action completes, in the order they were sent, possibly interleaved
+  # with action dispatches from other threads. Nested actions never deadlock
+  # with one another and a failure in a nested action will never affect the
+  # outer action.
+  #
+  # Nested actions can be called using the Agent reference from the enclosing
+  # scope or by passing the reference in as a "send" argument. Nested actions
+  # cannot be post using `self` from within the action block/proc/lambda; `self`
+  # in this context will not reference the Agent. The preferred method for
+  # dispatching nested actions is to pass the Agent as an argument. This allows
+  # Ruby to more effectively manage the closing scope.
+  #
+  # Prefer this:
+  #
+  # ```
+  # agent = Concurrent::Agent.new(0)
+  # agent.send(agent) do |value, this|
+  #   this.send {|v| v + 42 }
+  #   3.14
+  # end
+  # agent.value #=> 45.14
+  # ```
+  #
+  # Over this:
+  #
+  # ```
+  # agent = Concurrent::Agent.new(0)
+  # agent.send do |value|
+  #   agent.send {|v| v + 42 }
+  #   3.14
+  # end
+  # ```
+  #
+  # @!macro [new] agent_await_warning
   # 
-  #   bingo = Class.new{
-  #     def update(time, score)
-  #       puts "Bingo! [score: #{score}, time: #{time}]" if score >= 100
-  #     end
-  #   }.new
-  #   
-  #   score = Concurrent::Agent.new(0)
-  #   score.add_observer(bingo)
-  #   
-  #   score << proc{|current| sleep(0.1); current += 30 }
-  #   score << proc{|current| sleep(0.1); current += 30 }
-  #   score << proc{|current| sleep(0.1); current += 30 }
-  #   score << proc{|current| sleep(0.1); current += 30 }
-  #   
-  #   sleep(1)
-  #   #=> Bingo! [score: 120, time: 2013-07-22 21:26:08 -0400]
+  #   **NOTE** Never, *under any circumstances*, call any of the "await" methods
+  #   ({#await}, {#await_for}, {#await_for!}, and {#wait}) from within an action
+  #   block/proc/lambda. The call will block the Agent and will always fail.
+  #   Calling either {#await} or {#wait} (with a timeout of `nil`) will
+  #   hopelessly deadlock the Agent with no possibility of recovery.
```

Comment:
Probably. Interestingly, this is exactly the same as Clojure. I was surprised that Clojure took this approach but it is documented behavior.